### ISSUE TITLE
Unable to reconnect a disconnected connection.

### ISSUE THREAD

**Issue by starcraft66** [ammaraskar/pyCraft#146]
Hi, I didn't use to get these kinds of errors until recently when I updated pycraft. My disconnect/reconnect handling code looks like this:
```
    def handle_disconnect(self, json_data=""):
        self.logger.info('Disconnected.')
        if json_data:
            self.logger.info("Disconnect json data: %s", json_data)
        if self.connection.connected:
            self.logger.info("Forced a disconnection because the connection is still connected.")
            self.connection.disconnect(immediate=True)
        time.sleep(15)
        while not self.is_server_online():
            self.logger.info('Not reconnecting to server because it appears to be offline.')
            time.sleep(15)
        self.logger.info('Reconnecting.')
        self.connection.connect()

    def handle_disconnect_packet(self, disconnect_packet):
        self.handle_disconnect(disconnect_packet.json_data)
```
When my minecraft account gets disconnected for any reason, it is unable to reconnect and this gets printed to the console every time it tries to reconnect.
```
2019-10-05 23:25:09,469:bridge:INFO:Reconnecting.
2019-10-05 23:25:09,520:bridge:ERROR:A minecraft exception occured! 'NoneType' object has no attribute 'send':
Traceback (most recent call last):
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/connection.py", line 544, in run
    self._run()
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/connection.py", line 559, in _run
    while not self.interrupt and self.connection._pop_packet():
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/connection.py", line 297, in _pop_packet
    self._write_packet(self._outgoing_packet_queue.popleft())
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/connection.py", line 310, in _write_packet
    packet.write(self.socket)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/packets/packet.py", line 99, in write
    self._write_buffer(socket, packet_buffer, compression_threshold)
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/packets/packet.py", line 88, in _write_buffer
    VarInt.send(len(packet_buffer.get_writable()), socket)  # Packet Size
  File "/root/.local/share/virtualenvs/app-4PlAip0Q/lib/python3.7/site-packages/minecraft/networking/types/basic.py", line 165, in send
    socket.send(out)
AttributeError: 'NoneType' object has no attribute 'send'
```

**Comment by starcraft66** [comment]
I dug around the code a bit today and, visually, it doesn't look like it's possible for this to happen since it looks like we are trying to fire off some packets in the queue when the socket is set to `None` due to the client being disconnected.

I am speculating that this issue is arising due to a concurrency issue stemming from the fact that I am disconnecting and reconnecting the connection via an exception handler callback initiated by an active connection in the first place. I will spend more time debugging when I have some to spare.

**Comment by joodicator** [comment]
@starcraft66 In #202, you propose clearing the outgoing packet queue upon an `immediate` disconnect as a fix to this problem. Could you explain in more detail how this fixes it? Do you have a minimal example program that reproduces the error before the changes in #202 and does not have the error after them?

**Comment by joodicator** [comment]
For the record, this program did not reproduce this error:

```py
from minecraft.networking.connection import Connection
from minecraft.networking.packets import clientbound  

conn = Connection('localhost', username='User')

@conn.listener(clientbound.play.DisconnectPacket)
def handle_disconnect(packet):
    print('Reconnecting.')
    conn.connect()

conn.connect()
while conn.connected:
    (conn.networking_thread or conn.new_networking_thread).join()
```

**Comment by starcraft66** [comment]
@joodicator Sorry for the lack of clarification. The problem I observed happens when the connection to the server unexpectedly hangs up, such as when the server is extremely laggy or it just dies. When developing the fix, I could reproduce the original stack trace with a 100% success rate by starting a server and just `kill -9`ing the java process running it.

It appeared that there would still be packets to be sent in the outgoing queue that wouldn't be sent because of `immediate` being true. Those packets would then immediately attempt to send when reconnecting to the server before the socket was actually open.

**Comment by joodicator** [comment]
@starcraft66 If I understand correctly, the `DisconnectPacket` handler above was not involved in the error, but rather an exception handler not shown in the code snippet of your OP. Is that correct?

**Comment by starcraft66** [comment]
@joodicator Yes, that is correct. The exception handler was problematic and is what the PR fixes.

**Comment by joodicator** [comment]
@starcraft66 Would you be able to provide the code that reproduces this?

**Comment by joodicator** [comment]
@starcraft66 I think I've found how to reproduce it, so this isn't required anymore.

**Comment by starcraft66** [comment]
Alright great. Sorry for not coming up with a reproducer quckly, I'm busy with my finals for school right now.

**Comment by joodicator** [comment]
Now that we now the conditions in which this occurs, my analysis of the exact cause follows:

As starcraft66 correctly notes in #202, an exception occuring in the networking thread of a `Connection` calls the user-registered exception handlers and then, *regardless of what those handlers do*, terminates the current connection. This means that when an exception handler of the form
```py
@conn.exception_handler()
def handle_exception(exc, exc_info):
    conn.disconnect(immediate=True)
    conn.connect()
```
is present, the effect is that the following calls are made in quick succession:
```py
conn.disconnect(immediate=True)
conn.connect()
conn.disconnect(immediate=True)
```

# Problem 1
Given that this is happening, the expected result is that a connection attempt is started and then quickly terminated, leaving the connection in a disconnected state; however, that does not happen, which is the first problem. After the call to `connect` and before the second call to `disconnect`, the state of `conn` is as follows:
- `conn.networking_thread` is still running, but has `conn.networking.thread.interrupt == True`, meaning it will shortly exit.
- `conn.new_networking_thread` is running with `interrupt == False`, and waiting for the old thread to exit before it takes over to run the new connection,
- `conn.socket` is open and connected to the server, representing the connection that the new networking thread is to manage.

Looking at the code of `conn.disconnect`, we can see how the second call puts the connection into an invalid state:
https://github.com/ammaraskar/pyCraft/blob/252e60a6e2a3a8fb974779c4bce55530b62a8805/minecraft/networking/connection.py#L448-L470

After this is run, we have
- `conn.networking_thread.interrupt == True`, but
- `conn.new_networking_thread.interrupt = False`,
- `conn.connected == False`, and
- `conn.socket is None`.

The new networking thread was not terminated; once the old networking thread terminates, it tries to write to the nonexistent socket, resulting in the exception reported in this issue:
```
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/connection.py", line 591, in _run
    while not self.interrupt and self.connection._pop_packet():
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/connection.py", line 321, in _pop_packet
    self._write_packet(self._outgoing_packet_queue.popleft())
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/connection.py", line 334, in _write_packet
    packet.write(self.socket)
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/packets/packet.py", line 104, in write
    self._write_buffer(socket, packet_buffer, compression_threshold)
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/packets/packet.py", line 93, in _write_buffer
    VarInt.send(len(packet_buffer.get_writable()), socket)  # Packet Size
  File "/home/travis/build/ammaraskar/pyCraft/minecraft/networking/types/basic.py", line 177, in send
    socket.send(out)
AttributeError: 'NoneType' object has no attribute 'send'
```

The fix for this first error is to change `minecraft.Connection.disconnect` so that it properly handles the case when a new connection is starting and `new_networking_thread is not None`.

# Problem 2
The behaviour we desire when an exception occurs in a connection's networking thread is that:
- It is possible for the exception handler to examine the state of the connection as exactly as possible to when the exception occurred,
- After the exception handlers run, the old connection, which was running at the time the exception occurred, is terminated, but
- It is possible for an exception handler to choose to terminate the current connection and start a *new* connection that will *not* be automatically terminated.

An exception occuring in the `new_networking_thread` would indicate a highly abnormal condition, since the only code running then is
https://github.com/ammaraskar/pyCraft/blob/252e60a6e2a3a8fb974779c4bce55530b62a8805/minecraft/networking/connection.py#L568-L573
Therefore, I find it acceptable that a handler for such an exception will cause a `minecraft.exception.InvalidState` upon attempting to initiate a new connection, and assume in the following that the exception handler is running in the `networking_thread`, with the `new_networking_thread` identical to `None`.

My proposed fix is to terminate the connection only if a new connection has not been started - this can be detected by the presence of a `new_networking_thread` with `interrupt == False`. This will allow exception handlers such as the following to work as expected.
```py
@conn.exception_handler()
def handle_exception(exc, exc_info):
    conn.disconnect(immediate=True)
    conn.connect()
```

If this is implemented, then I see no need for the `disconnect_on_exception` flag proposed in #202, since in normal operation, the library should not rely on the user to leave the Connection in a valid state. If the user wants to hack the internals of pyCraft and prevent the default behaviour in response to an exception, it is possible to override the `Connection._handle_exception` method, with the leading underscore warning that this is not supported by the API.

**Comment by joodicator** [comment]
I have implemented the above proposals in commit 7372895, which should fix this issue. Feel free to comment if there is more.
